### PR TITLE
[WASI] WASI Native Thread

### PR DESCRIPTION
**Pull request by eee4017**
Instead of expose using WASI, these pthread API may be implemented with current thread proposal atomic support.

```
int pthread_mutex_init(pthread_mutex_t *mutex, const void *attr);
int pthread_mutex_lock(pthread_mutex_t *mutex);
int pthread_mutex_unlock(pthread_mutex_t *mutex);
int pthread_mutex_destroy(pthread_mutex_t *mutex);

int pthread_rwlock_init(pthread_rwlock_t *rwlock, const pthread_rwlockattr_t *attr);
int pthread_rwlock_rdlock(pthread_rwlock_t *rwlock);
int pthread_rwlock_tryrdlock(pthread_rwlock_t *rwlock);
int pthread_rwlock_trywrlock(pthread_rwlock_t *rwlock);
int pthread_rwlock_wrlock(pthread_rwlock_t *rwlock);
int pthread_rwlock_unlock();
int pthread_rwlock_destroy(pthread_rwlock_t *rwlock);

int pthread_cond_init(pthread_cond_t *cond, const void *attr);
int pthread_cond_wait(pthread_cond_t *cond, pthread_mutex_t *mutex);
int pthread_cond_timedwait(pthread_cond_t *cond, pthread_mutex_t *mutex, unsigned int useconds);
int pthread_cond_signal(pthread_cond_t *cond);
int pthread_cond_broadcast(pthread_cond_t *cond);
int pthread_cond_destroy(pthread_cond_t *cond);

int pthread_key_create(pthread_key_t *key, void (*destructor)(void *));
int pthread_setspecific(pthread_key_t key, const void *value);
void *pthread_getspecific(pthread_key_t key);
int pthread_key_delete(pthread_key_t key);
```

### CODE REVIEW

**Comment by eee4017** [comment]
If my understanding of the program structure is correct, the native thread could be designed as illustrated in the figure. The create_thread is called with a function pointer (fp) and some args (args).  Then, the Wasi Environ asks executor to create a thread. 

- The Wasi Environ should ask executor to create a thread. Thus, env should have access to VM, so Environ should initialize with the pointer of VM.
- The Executor should recover the state of the caller, e.g. `FunctionInstance` of caller, and maybe duplicate it. With `Table Instance`, the function pointer could be convert to function instance. Currently I'm not very sure how to do this.
  
Untitled Diagram drawio

**Comment by eee4017** [comment]
We could create `std::thread` with the current implementation now. However, I suspect that some internal racing causes failed execution. 

### How to examine the current implementation?

```
cd examples/thread/rust-wasi-native-thread
cargo build --target wasm32-wasi
wasmedge <path to WasmEdge>/examples/thread/rust-wasi-native-thread/target/wasm32-wasi/debug/rust-wasi-native-thread.wasm
```
The wasmedge panicked with this wasm file, and the situation of the panicking differs every run. Thus, I suspect the cause is some racing of the state in the wasmedge executor.


### How to debug with a thread sanitizer?

We could use thread sanitizer of clang to inspect the possible racing condition by simply adding CXX_FLAGS in global `CMakeLists.txt`.

```
set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=thread")
```

We could get the following report. There are mainly three kinds of racing
1. wasm memory `MemoryInstance::storeValue`: This should be fine since there are multiple threads accessing memory simultenously.
2. `controlInstr.cpp:88`: This is the `HostFunctionContext` that I add into wasmedge native thread that I will try to preserve the TableInst of the host function caller. The variable should be `thread_local`. Furthermore, If this induces panicking, It should happen when we try to recover the address at `createThreadWithFunctionAddress`. I have not found any read towards `HostFunctionContext.TableInst` under the racing condition. 

In summary, the cause of the panicking is unknown.


```
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race ../lib/executor/engine/controlInstr.cpp:88 in WasmEdge::Executor::Executor::runCallOp(WasmEdge::Runtime::StackManager&, WasmEdge::AST::Instruction casmEdge::AST::Instruction const*&, bool)
SUMMARY: ThreadSanitizer: data race ../lib/executor/engine/controlInstr.cpp:88 in WasmEdge::Executor::Executor::runCallOp(WasmEdge::Runtime::StackManager&, WasmEdge::AST::Instruction const&, WasmEdge::AST::Instruction const*&, bool)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179e9d) in std::enable_if<IsWasmNativeNumV<unsigned long>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned long, 8u>(unsigned long const&, unsigned int)
SUMMARY: ThreadSanitizer: data race ../include/runtime/instance/memory.h:305 in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 1u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race ../include/runtime/instance/memory.h:269 in std::enable_if<IsWasmNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::loadValue<unsigned int, 1u>(unsigned int&, unsigned int) const
SUMMARY: ThreadSanitizer: data race (/usr/local/bin/../lib/libwasmedge.so.0+0x3c76ef) in std::enable_if<IsWasmNativeNumV<unsigned long>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned long, 8u>(unsigned long const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/bin/../lib/libwasmedge.so.0+0x3c76ef) in std::enable_if<IsWasmNativeNumV<unsigned long>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned long, 8u>(unsigned long const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179e9d) in std::enable_if<IsWasmNativeNumV<unsigned long>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned long, 8u>(unsigned long const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179ccb) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/bin/../lib/libwasmedge.so.0+0x3c751d) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 4u>(unsigned int const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179e9d) in std::enable_if<IsWasmNativeNumV<unsigned long>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned long, 8u>(unsigned long const&, unsigned int)
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x179e9d) in std::enable_if<IsWasmNativeNumV<unsigned long>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned long, 8u>(unsigned long const&, unsigned int)
SUMMARY: ThreadSanitizer: data race ../include/runtime/instance/memory.h:269 in std::enable_if<IsWasmNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::loadValue<unsigned int, 1u>(unsigned int&, unsigned int) const
SUMMARY: ThreadSanitizer: data race (/usr/local/lib/wasmedge/libwasmedgePluginWasmEdgeThread.so+0x17a3c6) in std::enable_if<IsWasmNativeNumV<unsigned int>, cxx20::expected<void, WasmEdge::ErrCode> >::type WasmEdge::Runtime::Instance::MemoryInstance::storeValue<unsigned int, 1u>(unsigned int const&, unsigned int)
```

**Comment by harry900831** [comment]
> TL;DR
> The current implementation is incorrect. We should think more carefully to handle it. 


## Current Situation

The current implementation is as follows: 
Wasm file may import the host function `pthread_create` from the runtime. While calling `pthread_create`, we may get the function by `createThreadWithFunctionAddress` and use std::thread to run it. `createThreadWithFunctionAddress` should get the function instance that passes to pthread_create to run in multithread by function table and return a function that creates an independent `StackManager` and run `runFunction`.

As we can see above, the design may reuse the module, memory, table, global, ... and so on. Only the stack is independent. The original thought is that local variables may be stored in a stack and we can use memory to share between threads.  

But according to the wat file of the rust workload, it seems that even if our `my_thread` function didn't use any variable on memory, only thread local stuff, it may still manipulate on the memory or global instance. Moreover, all the threads may access the same memory offset on every memory operation in `my_thread`, which means that every `i32.load`, `i32.store`, `global.set`, and so on. access the same memory index. There's no mechanism to deal with it. Since the instruction manipulates on the same memory position, I believe that this is causing the runtime error.

```wat
(func $_ZN23rust_wasi_native_thread9my_thread17haa7178676c2b3f7cE (type $t4) (param $p0 i32) (result i32)
  (local $l1 i32) (local $l2 i32) (local $l3 i32) (local $l4 i32) (local $l5 i32) (local $l6 i32) (local $l7 i32) (local $l8 i32) (local $l9 i32)
  (local.set $l1
    (global.get $__stack_pointer))
  (local.set $l2
    (i32.const 16))
  (local.set $l3
    (i32.sub
      (local.get $l1)
      (local.get $l2)))
  (global.set $__stack_pointer
    (local.get $l3))
  (i32.store offset=4
    (local.get $l3)
    (local.get $p0))
  (local.set $l4
    (i32.const 0))
  (i32.store offset=8
    (local.get $l3)
    (local.get $l4))
  (local.set $l5
    (i32.const 0))
  (i32.store offset=12
    (local.get $l3)
    (local.get $l5))
  (local.set $l6
    (i32.load offset=12
      (local.get $l3)))
  (local.set $l7
    (call $_ZN4core3ptr8metadata18from_raw_parts_mut17h4c43a6e7e558ab23E
      (local.get $l6)))
  (local.set $l8
    (i32.const 16))
  (local.set $l9
    (i32.add
      (local.get $l3)
      (local.get $l8)))
  (global.set $__stack_pointer
    (local.get $l9))
  (return
    (local.get $l7)))
```

## Wasi-thread proposal
According to the [wasi-thread proposal](https://github.com/WebAssembly/wasi-threads), the design choice is actually: instance-per-thread, which is different than ours. Webassembly is not designed for multithread-per-instance. We may give a more comprehensive consideration of the design choice and implementation.

## Note 
- The type of thread id is unsigned in POSIX Threads, so the current rust workload may error if we link to libc.
- The method to get the table instance in `createThreadWithFunctionAddress` is inappropriate and may have bug.
- Since each thread share the memory, so we may need to lock the memory instruction to prevent from data race.

**Comment by hydai** [comment]
This will be replaced by #2505
